apps/sms/js/thread_list_ui.js

```diff
@@ -408,89 +502,92 @@ var ThreadListUI = {
   // please make sure url will also be revoked if new delete api remove threads
   // without calling removeThread in the future.
   delete: function thlui_delete() {
-    function performDeletion() {
-      /* jshint validthis: true */
+    var selection = this.selectionHandler;
 
-      var threadIdsToDelete = [],
-          messageIdsToDelete = [],
-          threadCountToDelete = 0,
-          selected = this.selectionHandler.selectedList;
+    var promise = new Promise(function(resolve) {
+      ThreadListUI.performDeletion(selection.selectedList);
+      resolve();
+    });
 
-      function exitEditMode() {
-        ThreadListUI.cancelEdit();
-        WaitingScreen.hide();
-      }
+    promise.then(function() {
+      ThreadListUI.onDeletionReadUnread({
+        l10nId: 'deleted',
+        l10nArgs: {
+          n: selection.selectedCount
+        }
+      });
+    });
+  },
```

Comment:
To be clearer: your current code does this:
* run performDeletion -- note that performDeletion is asynchronous, so it starts all getMessages and then returns
* when performDeletion returns, you call "resolve". That means the promise will be resolved.
* so when you run "promise.then()", this will be executed (asynchronously) just now.

And that's not really what we want :)

Remember we can't really undo deletion. So what we want to do is _delay_ deletion while the toast is displayed, but still make it appear as if it's really deleted. That's why I proposed a "UndoManager" when we discussed on IRC.

I think we need to do this UndoManager first, and only then apply it for delete/mark as read.

How about working on a undo for the draft saving operation? We already have a toast for this, and we can simply update it later. Then we can come back at this bug when we'll have a working UndoManager for draft saving.

What do you think?